slatedb/src/clock.rs

```diff
@@ -58,27 +58,56 @@ pub trait SystemClock: Debug + Send + Sync {
 /// specified duration has elapsed. This is to mimic Tokio's ticker behavior.
 pub struct SystemClockTicker<'a> {
     clock: &'a dyn SystemClock,
-    duration: Duration,
-    first_tick: bool,
+    duration: TimeDelta,
+    last_tick: AtomicI64,
 }
 
 impl<'a> SystemClockTicker<'a> {
     fn new(clock: &'a dyn SystemClock, duration: Duration) -> Self {
+        let duration = TimeDelta::from_std(duration).expect("duration is out of range");
         Self {
             clock,
             duration,
-            first_tick: true,
+            last_tick: AtomicI64::new(i64::MIN),
         }
     }
 
-    pub fn tick(&mut self) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
-        // Tokio's ticker ticks immediately when the first tick() is called.
-        // Let's emulate that behavior in our ticker.
-        if self.first_tick {
-            self.first_tick = false;
-            Box::pin(async {})
+    /// Returns a future that emits a signal every `duration` interval. The next tick is
+    /// calculated as last_tick + duration. The first tick will complete immediately.
+    /// This is to mimic Tokio's ticker behavior.
+    ///
+    /// If the clock advances more than the duration between `tick()` calls, the ticker
+    /// will tick immediately.
+    pub fn tick(&self) -> Pin<Box<dyn Future<Output = ()> + Send + '_>> {
+        Box::pin(async move {
+            let sleep_duration = self
+                .calc_duration()
+                .to_std()
+                .expect("duration is out of range");
+            self.clock.sleep(sleep_duration).await;
+            let now_dt = self.clock.now().timestamp_millis();
+            self.last_tick.store(now_dt, Ordering::SeqCst);
+        })
+    }
+
+    /// Calculates the duration until the next tick.
+    ///
+    /// The duration is calculated as `duration - (now - last_tick)`.
+    fn calc_duration(&self) -> TimeDelta {
+        let zero = TimeDelta::milliseconds(0);
+        let last_tick_ts = self.last_tick.load(Ordering::SeqCst);
+        if last_tick_ts == i64::MIN {
```

Comment:
can we just initialize `last_tick_ts` to `DateTime::MIN_UTC` and drop this case?

Reply:
Yep, good call.